lib/elixir/lib/task.ex

```diff
@@ -304,7 +307,14 @@ defmodule Task do
   again. To await the task's reply multiple times use `yield/2` instead.
   """
   @spec await(t, timeout) :: term | no_return
-  def await(%Task{ref: ref}=task, timeout \\ 5000) do
+  def await(task, timeout \\ 5000)
+
+  # TODO: Remove nil check in Elixir 1.3
+  def await(%Task{owner: owner}=task, _) when owner != nil and owner != self() do
+    raise_invalid_owner_error(task)
```

Comment:
When raising we want to provide the best information possible. If we raise in another function that is called in the tail of the function, the function will not be in the stacktrace. This occurs because of tail call optimisation removing the function from the stack because it is no longer needed.

To get around this either use a helper to create the exception or exception arguments, or ensure the function actually calling raise (or exit/1) is not called in the tail. I have applied such an improvement here: https://github.com/elixir-lang/elixir/commit/4c42568ddee209978dbe158eb825a599c8614d58.

This is very subtle point and may help you understand a future stacktrace that confuses you at first.